### PR TITLE
Expose topo_consul_lock_session_checks to allow customized consul session check

### PR DESCRIPTION
**Pull request by narcsfz**
Signed-off-by: crowu <y.wu4515@gmail.com>

## Description

When Consul creates lock, it needs to create a session for the lock. By default, session uses serfHealth check for liveness check: https://www.consul.io/api-docs/session, which is a node level healthy-check from gossip protocal. 

This means when the process restarted, if the node is still healthy, the session will remain alive up to 2*TTL (15sec) = 30 sec - during this period of time, other process cannot get the lock.

To address that, this PR exposes an optional flag to pass in customized session checks so the the session will expire automatically once the health check failed.

## Related Issue(s)

https://github.com/vitessio/vitess/issues/8297

## Checklist
- [x] Tests were added or are not required
- [x] Documentation was added or is not required

### CODE REVIEW

**Comment by narcsfz** [comment]
I see tests failed with error:
```
--- FAIL: TestConsulTopoWithChecks (0.24s)
Error:     testing.go:49: === checkKeyspace
Error:     server_flaky_test.go:302: OpenServer() failed: open /tmp/consul_auth_client_static_file.json160052189: no such file or directory
```

But it works on my IDE locally..

**Comment by narcsfz** [comment]
hm.. looks like [commenting out](https://github.com/vitessio/vitess/pull/8298/commits/1749fb5515c357714d55d2af7c7edba456bba7f1) my change fixes the test.. I will look into that

File: go/vt/topo/consultopo/lock.go [hunk 1]
```diff
@@ -86,14 +95,18 @@ func (s *Server) Lock(ctx context.Context, dirPath, contents string) (topo.LockD
...
 		// Failed to lock, give up our slot in locks map.
 		// Close the channel to unblock anyone else.
 		s.mu.Lock()
 		delete(s.locks, lockPath)
 		s.mu.Unlock()
 		close(li.done)
-
+		// Consul will return empty leaderCh with nil error if we cannot get lock before the timeout
+		// therefore we return a timeout error here
+		if lost == nil {
```
Comment (narcsfz) [inline go/vt/topo/consultopo/lock.go:107]: see https://github.com/hashicorp/consul/blob/master/api/lock.go#L185 - consul lock return nil error when the context timeout'd

**Comment by narcsfz** [comment]
ptal @deepthi @rafael

File: go/vt/topo/consultopo/lock.go [hunk 2]
```diff
@@ -51,11 +51,22 @@ func (s *Server) Lock(ctx context.Context, dirPath, contents string) (topo.LockD
 
 	lockPath := path.Join(s.root, dirPath, locksFilename)
 
-	// Build the lock structure.
-	l, err := s.client.LockOpts(&api.LockOptions{
+	lockOpts := &api.LockOptions{
 		Key:   lockPath,
 		Value: []byte(contents),
-	})
+		SessionOpts: &api.SessionEntry{
```
Comment (narcsfz) [inline go/vt/topo/consultopo/lock.go:57]: default settings from consul https://github.com/hashicorp/consul/blob/master/api/lock.go#L103-L107

**Comment by deepthi** [comment]
We'll need @rafael to review this (unless @guidoiaquinti or @ajm188 can).
In the meantime, can you squash the commits into a single one?

**Comment by narcsfz** [comment]
Thanks, I cleaned up the commits

**Review by guidoiaquinti — APPROVED** [review]
LGTM 👍 

As Rafael is PTO, @ajm188 can you 👀 as well?

**Comment by deepthi** [comment]
@narcsfz can you resolve the merge conflicts and push again?

**Comment by narcsfz** [comment]
> @narcsfz can you resolve the merge conflicts and push again?

yep, rebased and resolved conflicts

**Comment by rafael** [comment]
LGTM

**Pull request merged**